lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp

```diff
@@ -0,0 +1,313 @@
+//===------ YieldOnceCheck.cpp - Check usage of yields in accessors  ------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "yield-once-check"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/DiagnosticsSIL.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/Stmt.h"
+#include "swift/SIL/BasicBlockUtils.h"
+#include "swift/SIL/Dominance.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "llvm/ADT/DenseSet.h"
+
+using namespace swift;
+
+namespace {
+
+class YieldOnceCheck : public SILFunctionTransform {
+
+  template <typename... T, typename... U>
+  static InFlightDiagnostic diagnose(ASTContext &Context, SourceLoc loc,
+                                     Diag<T...> diag, U &&... args) {
+    return Context.Diags.diagnose(loc, diag, std::forward<U>(args)...);
+  }
+
+  /// Data-flow analysis state that is associated with basic blocks.
+  struct BBState {
+
+    /// Indicates whether a basic block is encountered before seeing a yield
+    /// (BeforeYield) or after seeing a yield (AfterYield), or in both states
+    /// (Conflict). This enum is a semi-lattice where Conflict is the top and
+    /// the merge of BeforeYield and AfterYield states is Conflict.
+    enum YieldState { BeforeYield, AfterYield, Conflict } yieldState;
+
+  private:
+    // The following state is maintained for emitting diagnostics.
+
+    // For AfterYield and Conflict states, this field records the yield
```

Comment:
This should be a doxygen comment. (doxygen comment == '///'). We really need to get some sort of doxygen producing job up. It will make it easier to make sure this works well.